metricflow-semantics/metricflow_semantics/time/dateutil_adjuster.py

```
@@ -0,0 +1,114 @@
+
+    * `relativedelta` will not change weekday if already at the given weekday, even with a Nth parameter.
+    * `relativedelta` will automatically handle day values exceeding for months with fewer days.
+    """
+
+    def _relative_delta_for_window(self, time_granularity: TimeGranularity, count: int) -> relativedelta:
+        """Relative-delta to cover time windows specified at different grains."""
+        if time_granularity is TimeGranularity.DAY:
+            return relativedelta(days=count)
+        elif time_granularity is TimeGranularity.WEEK:
+            return relativedelta(weeks=count)
+        elif time_granularity is TimeGranularity.MONTH:
+            return relativedelta(months=count)
+        elif time_granularity is TimeGranularity.QUARTER:
+            return relativedelta(months=count * 3)
+        elif time_granularity is TimeGranularity.YEAR:
+            return relativedelta(years=count)
+        else:
+            assert_values_exhausted(time_granularity)
+
+    @override
+    def expand_time_constraint_to_fill_granularity(
+        self, time_constraint: TimeRangeConstraint, granularity: TimeGranularity
+    ) -> TimeRangeConstraint:
+        adjusted_start = self.adjust_to_start_of_period(granularity, time_constraint.start_time)
+        adjusted_end = self.adjust_to_end_of_period(granularity, time_constraint.end_time)
+
+        if adjusted_start < TimeRangeConstraint.ALL_TIME_BEGIN():
+            adjusted_start = TimeRangeConstraint.ALL_TIME_BEGIN()
+        if adjusted_end > TimeRangeConstraint.ALL_TIME_END():
+            adjusted_end = TimeRangeConstraint.ALL_TIME_END()
+
+        return TimeRangeConstraint(start_time=adjusted_start, end_time=adjusted_end)
+
+    @override
+    def adjust_to_start_of_period(
+        self, time_granularity: TimeGranularity, date_to_adjust: datetime.datetime
+    ) -> datetime.datetime:
+        if time_granularity is TimeGranularity.DAY:
+            return date_to_adjust
```

This assumes all input has a minimum granularity of daily and silently returns whatever is passed in. The Pandas implementation raises an exception in this case, which might be preferable for testing scenarios.

I'm fine with either approach since we're going to update this within the next couple of weeks, but @courtneyholcomb is going to be doing the work on opening up granularity so maybe check in with her about what she prefers before merging.

---

Pinged her, but preferring to fix later if there are issues to unblock merging of the stack.